Proposal: Add DPM Trace source-aware debugging (#327)
canton-foundation/canton-dev-fund: proposals/dpm-trace-source-aware-debugging.md

> # Development Fund Proposal: DPM Trace and Source-Aware Debugging

Reviewer: The title should be changed because it is misleading since "Source aware debugging" is not provided.  
It is more aptly described as interactive parser or visualizer.

Author: I completely agree.

Since we will separate out “Debug Info” generation into new proposal, this can stay as interactive inspector/visualization of the trace as suggested. We’ll submit the separate Debug Info generation proposal soon - happy to discuss whether you’d prefer us working on both in parallel or sequentially (this tx viz proposal first, then debug info as a follow on). The Draft PR for it is [Debug Info Generation PR](https://github.com/canton-foundation/canton-dev-fund/pull/382).

Reviewer 2: Agreed too, I think that weakens the synergy I was seeing with #323 unfortunately

Author: @monsieurleberre thanks for the comment!

>Agreed too, I think that weakens the synergy I was seeing with https://github.com/canton-foundation/canton-dev-fund/pull/323 unfortunately

Yes, exactly. But we are working on a separate proposal that will target debug info generation from the damlc. We will present how it can be used in both debugging tooling as well as in profiling/test coverage.

I would add that we do have experience with it  -- 10+ years of dealing with debug info related issues in compiler infrastructures such as LLVM, and currently we are in the process of adding debug info generation to the official compiler for Solidity - solc - via collaboration w Argot / Ethereum Foundation.

Author: Btw, here Is a draft https://github.com/walnuthq/canton-dev-fund/blob/pre/debug-info-generation/proposals/daml-versioned-debug-info-metadata-draft.md. It is in a very early stage, but we are working on it.

---

## Table of Contents

- [Abstract](#abstract)
- [Specification](#specification)
  - [1. Objective](#1-objective)
  - [2. Motivation](#2-motivation)
  - [3. Implementation Mechanics](#3-implementation-mechanics)
  - [4. Architectural Alignment](#4-architectural-alignment)
  - [5. Backward Compatibility](#5-backward-compatibility)
- [Proof of Concept Implementation](#proof-of-concept-implementation)
- [Milestones and Deliverables](#milestones-and-deliverables)
  - [Milestone 1: Transaction Trace CLI](#milestone-1-transaction-trace-cli)
  - [Milestone 2: Trace Bundles](#milestone-2-trace-bundles)
  - [Milestone 3: Interactive Debugger](#milestone-3-interactive-debugger)
  - [Milestone 4: Debug Information Format and Compiler Integration](#milestone-4-debug-information-format-and-compiler-integration)
  - [Milestone 5: Adoption and Ecosystem Validation](#milestone-5-adoption-and-ecosystem-validation)
- [Acceptance Criteria](#acceptance-criteria)
- [Potential Follow-Ons](#potential-follow-ons)
- [Funding](#funding)
- [Co-Marketing](#co-marketing)
- [Rationale](#rationale)
- [About the Team](#about-the-team)
- [References](#references)

---

## Abstract

> We propose to build `dpm trace`, an open-source DPM plugin for inspecting and debugging Canton transactions.

Reviewer: Is it 'debugging Canton transactions' or 'visualizing Canton transactions'?  

BTW, we think there is a lot of value to being able to visualize the transactions but we don't want the proposal to claim more than it provides.

Author: We will apply this in our proposal, thanks!

> A developer should be able to start from an already committed update id, run a single CLI command, and get a readable transaction trace.

Reviewer: This proposal would  offer a lot of value if there is a  way to visually assess the situation when transaction doesn’t go according to expectation.  

Why can't the visualization be done with a Prepare transaction request?  Then you can perhaps do a diff if that makes sense.  

A diff between different instances (i.e., different update_id's) of the same business transaction may also make sense.

A diff between a Prepared transaction and the failed execution of that transaction would be helpful, especially if it incorporated log analysis across validators to identify a log message that can point to the root cause.

Author: Thank you, this is a good direction. We will update the proposal so `dpm trace` is not only a committed-update viewer, but also supports prepare and compare (diff) workflows.

Our understanding is:

  - committed transaction inspection uses `/v2/updates/update-by-id`
  - prepared transaction inspection uses `/v2/interactive-submission/prepare`
  - failed transaction analysis uses `/v2/commands/completions`

This lets the tool compare:

  - a prepared transaction against a successful committed one
  - two committed transactions that represent the same business operation
  - a prepared transaction against a failed completion

**For the validator/participant logs**: It is very nice idea, but how would you recommend obtaining them? Is there an authorized API or structured export for these logs, and is the format stable enough?

Reviewer: The logs should be available from the validator the dApp is running on.

There may be some operational security aspects to getting the logs for a MainNet (perhaps TestNet) validator but should be available for LocalNet / DevNet.

> From there, the tool adds interactive inspection and source-aware debugging workflows. To make that reliable, we also propose compiler-generated debug metadata so transaction traces and debugger steps can link back to Daml source.

Reviewer: Please remove the "propose compiler-generated debug metadata so transaction traces and debugger steps can link back to Daml source."  This is a challenging task and should be a second proposal with detailed technical content (i.e., explain the changes to the runtime to make debugger operations possible, such as breakpoints) to assess the approach.  The current transaction visualization is sufficient a a proposal.

Author: Agreed. As we mentioned above, we are on the same page!

We still believe that rich debug metadata generation is important long term, but it deserves a separate technical proposal (with maintainers involved of course).

This is an area where Walnut has relevant experience: our team members have worked for years on debug information in LLVM and on improving debugging quality for optimized C/C++ and Rust code.
We are also official maintainers of ETHDebug support in `solc`, and actively working on the implementation of it in the compiler.

But for this proposal, we agree that the current transaction inspection/visualization work is already valuable and should stand on its own.

---

## Specification

### 1. Objective

> Build the core tooling needed for source-aware Canton transaction debugging.

Reviewer: Please do a global replace to remove the term 'debuggin' or 'debugger' as it can mislead the casual reader.

- Add `dpm trace <update-id>` so developers can inspect already committed updates from the CLI.
- Add an interactive debugger so developers can inspect committed updates and replay captured trace bundles.

We are open to adjusting the details with Canton developers, Daml/Canton maintainers, and the Tech & Ops Committee.

### 2. Motivation

The Canton developer survey mentions Tenderly-like or Foundry-like tooling. We agree with the direction, but a naive 1:1 copy does not fit Canton's unique architecture.

We analysed the current Canton stack and found useful primitives that already exist:

- Transaction/update trace data is available through participant Ledger APIs, including the JSON Ledger API.
- Daml Script and IDE flows can produce useful local traces while tests or scripts are running.
- Participants already upload DARs and use package vetting to control which packages they are willing to process.

The gaps are:

> - There is no CLI command for inspecting the trace of a transaction / update.

Reviewer: Please add functionality to close the gaps of:

- Diff between Prepared and actual successful tx
- Diff between Prepared and failed tx that incorporates log analysis to identify the root cause
- Diff between two txs of the same business operation

- There is no interactive debugger workflow for committed updates.
- There is no interactive debugger for local Daml Script / test workflows
- Current package metadata can help decode Daml values, templates, choices, and fields, but it does not give debuggers reliable source locations or expression-level information.
- Daml/Canton tooling lacks rich, portable debug-info for reliable source-level debugging, including expression spans, execution step IDs, and consistent variable/value visibility.

This proposal focuses on those gaps.

### 3. Implementation Mechanics

The work is split into three core components.

#### DPM Trace CLI

Add a DPM plugin command:

```bash

> dpm trace <update-id> --participant <participant-url> --read-as <party>

Reviewer: Might recommend changing `--participant `to `--submitter`

Author: It may work as well. Thanks.

```

> The command connects to an authorized participant endpoint, fetches the participant-visible update, decodes it, and renders a readable trace tree with decoded templates, choices, arguments, payloads, and return values where visible.

Reviewer: I think only the submitter is authorized.  You should read the result from the completions stream which will provide event information for failed transactions.

It uses existing APIs:

- `UpdateService.GetUpdateById`
- JSON Ledger API `/v2/updates/update-by-id`
- `PackageService.GetPackage` where package metadata is available
- and optional event/query APIs for related visible contract payloads

> #### Interactive Debugger

Reviewer: The proposal is missing a key description of the UX for how the transaction is visualized.  It mentions a REPL but that doesn't explain how a user will easily navigate and easily comprehend the transaction.  Without an intuitive visualization, a moderate transaction tree will be challenging to navigate.

Can you please update the proposal with the UX?

Author: Agreed. We will update the proposal to describe the visualization UX more clearly.

I recorded a short CLI demo (please check it) of how we envision the transaction visualizer. It's a small example, but it shows the UX direction we want:

* the tool opens an already committed update from the participant-visible projection
* it shows the update as a navigable transaction tree
* n / p move through the trace event by event, with the selected event highlighted in the tree
* vars shows the decoded values relevant to that event, such as choice arguments, actors, witnesses, contract id, and created payload.

The proposal isn't just to "print a trace once." The intent is an interactive visualizer that lets a developer move around the update, inspect the key event data, and understand how the transaction tree changed state. For larger transactions, we'd extend the same model with search/filtering by template, choice, party, contract id, event type, and package id.

This is meant to lay the V1 foundation. Once users start playing with it, we will work with them, collect feedback and iterate. 

The example is a happy path (successful transaction); we're glad to put together a failed example or show what a diff might look like if there's interest.

Does this UX match your expectations? Any feedback welcome, and we're happy to iterate.

https://github.com/user-attachments/assets/98161808-1c5d-49b5-993f-7bcc716b3e75

Reviewer: Thanks for the update.  We have a customer that has 500 fields on its core template which is quite unusual but there will be a subset of customers with this large set of fields.   Indicating how this can be accommodated would be helpful.

Author: Sure, that makes sense. We will indicate this in the updated proposal. For large templates, the visualizer should not dump all fields at once. It can use the same general approach as tools like GDB/LLDB: show a compact summary first, collapse or page long field lists, ask/prompt the user before continuing when output exceeds the terminal window, and provide field search/filtering so users can jump directly to the fields they care about.

> Build an interactive debugger for transactions/updates.

Reviewer: ```suggestion
Build an interactive visualizer for transactions completions.
```

We propose adding an `--interactive` flag to `dpm trace` command to invoke the debugger, such as:

```plain text
dpm trace <update-id> --interactive
```

> This opens a REPL-style trace inspector for the committed update. The debugger can step through the trace, inspect decoded values and state changes, jump to selected trace events or source locations, and show source links where debug metadata is available. If debug metadata is missing, it still shows the event tree and visible ledger data. The richer the compiler-generated debug information is, the better the debugger experience becomes.

Reviewer: A recommended path for getting symbol information is:

- LF already has span information
- it is available by combining damlc inspect and damlc package-manifest
- they also need source file paths, they can get them from the daml.yaml file 
- IDE is doing all that already, they can take inspiration
- this gives them integration into multi-package from the get-go

Author: Thank you! We agree the first version should reuse existing LF/package/project metadata instead of defining new debug-info format.

Additionally, we propose adding two commands, `bundle` and `replay`, for replay-based debugging of Canton transactions/updates. `bundle` saves the participant-visible trace and available context as a portable JSON artifact. `replay` opens that artifact later without reconnecting to the original participant.

The bundle format decouples capture from inspection, enabling a Sentry-like architecture: an authorised backend captures bundles from a participant, while a Canton Explorer or future Tenderly-like tooling serves them to users via a web UI. `replay` is the local CLI consumer of the same format.

The commands will look as follows:
<empty-block/>
```bash

> dpm trace bundle <update-id> --participant <participant-url> --read-as <party> --out <trace-bundle>

Reviewer: How is authorization enabled for the LAPI calls?

Author: Our current assumption is that authorization should follow the normal Ledger API / JSON API model: the user provides the authorized endpoint, an access token, and the relevant party context.
  
For example:

```bash
dpm trace <update-id> \
     --submitter <url> \
     --read-as <party> \
     --access-token-file <token-file>
 ```
 
The tool would only show the projection available under that authorization context. Does this match how you would expect the tool to authenticate, or would you recommend a different approach?

Reviewer: Maybe allow for passing in the token as an env var

dpm trace replay <trace-bundle> --interactive
```

#### Rich Compiler-generated Debug Information

Current DAR metadata helps with package ids, modules, templates, choices, fields, and typed values. That is enough for decoding many traces, but it is not enough for robust source-level debugging.

> We propose defining a versioned Canton debug information format for Daml contracts. The format can be embedded in a DAR or emitted as a sidecar build artifact.

Reviewer: A versioned Canton debug information format for Daml contracts is technically challenging.  It should be a separate CIP and/or proposal.  Please remove.

Author: As mentioned above, will remove it.

Initial data:

- package id
- package name and version
- source root
- source file hashes
- module spans
- template spans
- choice spans
- expression spans where feasible
- deterministic debug step IDs that map source spans to Daml-LF/runtime execution points

> Later extensions:

Reviewer: Please remove.

- Debug-info output includes deterministic debug step IDs for supported source spans.
- The format is documented and versioned.
- The design is shared with Daml/Canton maintainers.

### Milestone 5: Adoption and Ecosystem Validation

**Estimated Delivery:** 4 weeks after Milestone 4 acceptance
**Focus:** Prove that the tooling is usable by Canton developers outside Walnut.

**Deliverables / Value Metrics:**

- Public release of the `dpm trace` plugin with installation instructions.
- Getting-started guide covering local Canton, authorized remote participant usage, trace bundles, and interactive debugging.
- At least one short demo video or recorded walkthrough.
- Outreach to Canton ecosystem developers, Daml/Canton maintainers, and Tech & Ops reviewers for hands-on testing.
- Feedback collection from at least three independent Canton developers or teams.
- Fixes or follow-up issues for adoption-blocking bugs found during the adoption window.
- Short adoption report summarizing who tested the tool, what workflows were tested, what worked, what did not, and recommended next steps.

**Acceptance Criteria:**

> - At least three independent Canton developers or teams run the trace workflow against a local project, sample project, or authorized participant endpoint, confirmed by written feedback, public issue/PR activity, or other verifiable evidence.

Reviewer: ```suggestion
- At least three independent organizations run the trace workflow against a dApp project and confirm that it sped up their development process and how it did so.  The confirmation can be by written feedback, public issue/PR activity, or other verifiable evidence.
```

> - At least two independent users test trace bundles or the interactive debugger, confirmed by written feedback, public issue/PR activity, or other verifiable evidence.

Reviewer: ```suggestion
- At least two independent organizations test trace bundles or the interactive debugger, confirmed by written feedback, public issue/PR activity, or other verifiable evidence.
```

- Self-hosted/private instance for institutions that cannot publish source externally.
- Storage of DARs where permitted, source files, package ids, debug-info artifacts, template/choice/field metadata, source spans, and optional repository metadata.
- `dpm trace` lookup by package id to make historical traces source-aware.

Open questions to resolve before this becomes its own proposal:

- Reproducibility of `damlc` builds across SDK versions, OS, and build environments, and a verification status model that handles partial matches.
- Threat model: who can publish, anti-squatting on package ids, source poisoning, and spam control.
- Licensing of registry data, separate from the Apache-2.0 code license.
- How pre-existing DARs without debug info are represented.

### Other Likely Follow-Ons

- A web UI on top of the trace bundle format and debug-info format.
- A DAP-compatible adapter and VS Code extension for the interactive debugger.
- Simulation and what-if tooling on top of replay bundles.
- Profiling and coverage extensions built on top of the debug-info format.

---

## Funding

> **Total Funding Request:** 1,900,000 Canton Coin

Reviewer: The funding needs to be structured so that 50% is allocated based on value added to the ecosystem.  For example, based on adoption.

The funding scheme does not match that.

I think there is also a request to add some sort of effort estimate against the funding as well.

Also, can you add in the diff capability?

Author: Yes, we will do it. Thanks!
